src/graph/LdapAuthenticator.cpp

```diff
@@ -0,0 +1,361 @@
+    }
+
+    if (FLAGS_ldap_scheme.compare("ldap")) {
+        return Status::Error("LDAP authentication ldap_scheme only support \"ldap\".");
+    }
+
+    if (FLAGS_ldap_port == 0) {
+        FLAGS_ldap_port = ldapPort_;
+    }
+
+    // Simple bind mode
+    if (!FLAGS_ldap_prefix.empty() || !FLAGS_ldap_suffix.empty()) {
+        if (!FLAGS_ldap_basedn.empty() ||
+            !FLAGS_ldap_binddn.empty() ||
+            !FLAGS_ldap_bindpasswd.empty() ||
+            !FLAGS_ldap_searchattribute.empty() ||
+            !FLAGS_ldap_searchfilter.empty()) {
+            return Status::Error("LDAP authentication direct bind mode and search bind mode"
+                                 " parameters cannot be mixed.");
+        }
+    // Here search bind mode must be set FLAGS_ldap_basedn
+    } else if (FLAGS_ldap_basedn.empty()) {
+        return Status::Error("LDAP authentication requires argument \"ldap_prefix\", "
+                             "\"ldap_suffix\", or \"ldap_basedn\" to be set.");
+    }
+
+    /**
+     * Search bind mode can either use FLAGS_ldap_searchattribute or
+     * FLAGS_ldap_searchfilter, but can't use both. FLAGS_Ldap_searchattribute
+     * default is "uid". FLAGS_Ldap_searchfilter is more flexible search filtes
+     * than FLAGS_Ldap_searchattribute.
+     */
+    if (!FLAGS_ldap_searchattribute.empty() && !FLAGS_ldap_searchfilter.empty()) {
+        return Status::Error("LDAP authentication cannot use ldap_searchattribute "
+                             "together with ldap_searchfilter.");
+    }
+
+    return Status::OK();
+}
+
+
+Status LdapAuthenticator::initLDAPConnection() {
+    int ldap_version = LDAP_VERSION3;
+
+    // Ldap_create interface is only available in versions higher than 2.4.49
+    /*
+    int ret  = ldap_create(&ldap_);
+    if (ret != LDAP_SUCCESS) {
+        return Status::Error("Init LDAP failed.");
+    }
+    ldap_->ld_options.ldo_defport = FLAGS_ldap_port;
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_HOST_NAME, FLAGS_ldap_server);
+    if (ret != LDAP_SUCCESS) {
+        ldap_ld_free(ldsp_, 1, NULL, NULL);
+        return Status::Error("Init LDAP failed.");
+    }
+    */
+    // ldap_init is deprecated, so use ldap_initialize
+
+    // FLAGS_ldap_server is comma separated list of hosts
+    auto ldap_server = folly::trimWhitespace(FLAGS_ldap_server);
+    if (ldap_server.empty()) {
+        return Status::Error("LDAP authentication ldap_server is illegal.");
+    }
+    const char *host = ldap_server.str().c_str();
+    std::string uris("");
+    do {
+        auto size = std::strcspn(host, ",");
+        if (uris.length() > 0) {
+            uris.append(" ");
+        }
+        uris.append(FLAGS_ldap_scheme);
+        uris.append("://");
+        uris.append(host, size);
+        uris += folly::stringPrintf(":%d", FLAGS_ldap_port);
+        host += size;
+        if (*host == ',') {
+            host++;
+        }
+        while (*host == ' ') {
+           host++;
+        }
```

Comment:
Use folly::split? A little difficult to understand this block

Reply:
Good catch.